Java/core/src/main/java/com/amazon/randomcutforest/store/NodeStore.java

```diff
@@ -34,40 +39,78 @@
     public final double[] cutValue;
     public final int[] mass;
+    public final int[] leafPointIndex;
+    public IndexManager freeNodeManager;
+    public IndexManager freeLeafManager;
 
     /**
      * Create a new NodeStore with the given capacity.
      * 
      * @param capacity The maximum number of Nodes whose data can be stored.
      */
     public NodeStore(int capacity) {
-        super(capacity);
-        parentIndex = new int[capacity];
+        this.capacity = capacity;
+        freeNodeManager = new IndexManager(capacity);
+        freeLeafManager = new IndexManager(capacity + 1);
+        parentIndex = new int[2 * capacity + 1];
+        mass = new int[2 * capacity + 1];
         leftIndex = new int[capacity];
         rightIndex = new int[capacity];
         cutDimension = new int[capacity];
         cutValue = new double[capacity];
-        mass = new int[capacity];
+        leafPointIndex = new int[capacity + 1];
+        Arrays.fill(parentIndex, NULL);
+        Arrays.fill(leftIndex, NULL);
+        Arrays.fill(rightIndex, NULL);
+        Arrays.fill(leafPointIndex, PointStore.INFEASIBLE_POINTSTORE_INDEX);
     }
 
-    public NodeStore(int[] parentIndex, int[] leftIndex, int[] rightIndex, int[] cutDimension, double[] cutValue,
-            int[] mass, int[] freeIndexes, int freeIndexPointer) {
+    public NodeStore(int capacity, int[] leftIndex, int[] rightIndex, int[] cutDimension, double[] cutValue,
+            int[] leafMass, int[] leafPointIndex, int[] freeNodeIndexes, int freeNodeIndexPointer,
+            int[] freeLeafIndexes, int freeLeafIndexPointer) {
         // TODO validations
-        super(freeIndexes, freeIndexPointer);
-        this.parentIndex = parentIndex;
+        this.capacity = capacity;
+        this.freeNodeManager = new IndexManager(capacity, freeNodeIndexes, freeNodeIndexPointer);
+        this.freeLeafManager = new IndexManager(capacity + 1, freeLeafIndexes, freeLeafIndexPointer);
+        this.parentIndex = getParentIndex(leftIndex, rightIndex);
         this.leftIndex = leftIndex;
         this.rightIndex = rightIndex;
         this.cutDimension = cutDimension;
         this.cutValue = cutValue;
-        this.mass = mass;
+        this.mass = new int[2 * capacity + 1];
+        // copy leaf mass to the later half; if mass is not null
+        if (leafMass != null) {
+            validateInternalState(leafPointIndex != null, " incorrect state for needing samplers");
+            System.arraycopy(leafMass, 0, this.mass, capacity, capacity + 1);
+            this.leafPointIndex = leafPointIndex;
+        } else {
+            this.leafPointIndex = new int[capacity + 1];
+            Arrays.fill(this.leafPointIndex, PointStore.INFEASIBLE_POINTSTORE_INDEX);
+        }
+        if (leafMass != null) {
+            for (int i = 0; i < capacity; i++) {
+                if (parentIndex[i] == NULL) {
+                    rebuildMass(i);
+                }
+            }
```

Comment:
Can we move this inside the previous if-block?

Also, since `rebuildMass` is recursive, would it suffice to call `rebuildMass` once on the root node?

Reply:
The node store does not know the root :) But that is an issue we perhaps should fix ...

Reply:
Kept this the same for now, maybe this would be resolved in a reorganization of code.